Easy Coding standard added

TomasVotruba opened the pull request:
Tool to check sniffs and fixers with simple neon config: https://github.com/Symplify/EasyCodingStandard

Coding standards:

- helps write error-proof an better code, actually doesn't force you to extra work, but writes it for you :+1: , 
- makes **code review easier to check and focus on business logic and flow, that robots can't check**
- and **allows practically anyone to contribute** without studying the way our code is structured


### Also  

- bumped SendGrid from 4.0 to 5.4, due to old dependencies
- setup checkers to validate standard

### Suggestions

- use PSR-2 spaces, lot checkers depends on them, so It

TomasVotruba commented:
**Ready to merge. Please check if you can spot any breaking changes.**

It took 1-2 days, but was easy thanks to very nicely structured code. Good job :clap: 

I'd also suggest using spaces as PSR-2, because many sniffs and checkers that might helps us depend on it. And it has pretty much PSR-2 style already anyway.

Btw, if you have some tip for new checker, let me know!

filipsuk commented on app/Modules/Newsletter/Model/NewsletterService.php:
```diff
@@ -143,24 +136,25 @@ public function sendNewsletters(array $usersNewslettersIds)
+			$from = new Email('Eventigo.cz', $userNewsletter->from);
+
+			$subject = $userNewsletter->subject;
+			$content = new SendGrid\Content('text/plain', $userNewsletter->content); // @todo: or html?
```
Have you ever seen eventigo newsletter? :) of course it's HTML

TomasVotruba replied:
Ok, I've onyl upgraded sendgrid from 4 to 5 and never used it, so just asking to be sure :)

filipsuk commented on app/Modules/Newsletter/Model/NewsletterService.php:
```diff
@@ -143,24 +136,25 @@ public function sendNewsletters(array $usersNewslettersIds)
+
+			$subject = $userNewsletter->subject;
+			$content = new SendGrid\Content('text/plain', $userNewsletter->content); // @todo: or html?
+			//	->setCategory('newsletter') // what is this for
```
category is for statistical purposes, e.g. you want to measure open rate of newsletters excluding login emails

TomasVotruba replied:
I see. 

Do you have idea where to put it in SendGrid 5.0?
I found this, but it's very blind without tests: https://github.com/sendgrid/sendgrid-php/blob/master/examples/categories/categories.php#L14

I'd appreciate help with this, never used it.

filipsuk commented on app/Modules/Front/Components/Tags/Tags.php:
```diff
@@ -36,12 +36,13 @@ class Tags extends BaseControl
+        UserTagModel $userTagModel,
+        TagGroupModel $tagGroupModel,
+        User $user
+) {
```
missing indent?

filipsuk commented on app/Modules/Front/Components/SubscriptionTags/SubscriptionTags.php:
```diff
@@ -33,12 +33,14 @@ class SubscriptionTags extends Subscription
+        TagModel $tagModel,
+        UserTagModel $userTagModel,
+        TagGroupModel $tagGroupModel
+) {
```
missing indent?

filipsuk commented on app/Modules/Email/Presenters/EmailPresenter.php:
```diff
@@ -3,16 +3,15 @@
 /**
- * EmailPresenter is used for rendering different types of emails.
- * @package App\Modules\Email\Presenters
+ * EqmailPresenter is used for rendering different types of emails.
```
typo EqmailPresenter

filipsuk commented on app/Modules/Email/Presenters/EmailPresenter.php:
```diff
@@ -45,8 +44,7 @@ class EmailPresenter extends BasePresenter
 	public function renderLogin($token)
 	{
-		$this->template->getLatte()->setLoader(new StringLoader);
+		$this->template->getLatte()->setLoader(new StringLoader); // @todo: what is this for?
```
don't remember, probably to be able to pass string to setFile() on next line?

TomasVotruba replied:
Thanks, that might be it.
I think we find out for use, when I add unit tests.

filipsuk commented on app/Modules/Core/Components/AbstractBaseControl.php:
```diff
@@ -31,7 +31,7 @@ protected function createTemplate(): Template
 		}
 
-		$template->addFilter('datetime', function (DateTime $a, DateTime $b = null) {
+		$template->addFilter('datetime', function (DateTime $a, ?DateTime $b = null) {
```
problem is that eventigo.cz is running on rosti.cz hosting and they don't support 7.1 yet :/ so this can't be merged until they support it or we change hosting

filipsuk commented on app/Modules/Core/Utils/DateTime.php:
```diff
@@ -53,10 +53,12 @@
 	 */
 	public static function max(...$dateTimes)
 	{
+        $dateTimes = array_filter($dateTimes); // filter out NULLs
```
spaces instead of tabs - did your IDE load .editorconfig? probably in more files

filipsuk commented on app/Modules/Core/Components/AbstractBaseControl.php:
```diff
@@ -10,49 +10,51 @@
-	{
-		$this->translator = $translator;
-	}
+    public function __construct(Translator $translator)
```
spaces instead of tabs

TomasVotruba replied:
Good catch! `PhpCsFixer\Fixer\Whitespace\IndentationTypeFixer` will help with that

filipsuk commented on app/Modules/Admin/Model/EventService.php:
```diff
@@ -17,7 +17,7 @@
 	 * @var string
 	 */
-	const PLATFORM_FACEBOOK = 'facebook';
+	public const PLATFORM_FACEBOOK = 'facebook';
```
again, problem with our PHP 7.0 hosting

filipsuk commented:
I tried some things on localhost and fixed the exceptions. I feel like you don't have the project running on your machine and do blind programming? :) shame we don't have tests

TomasVotruba commented:
@filipsuk Thanks a lot for checking this!

I'll check the fixer and maybe write some tests.

> I feel like you don't have the project running on your machine and do blind programming?

I have just partially checking only few sites. I'm not sure what to do, because the page looks completely different from the one online. Maybe we could do some coffee meetup, so you could integrate me?

jakubzaba commented:
@TomasVotruba Good job 👍. PHP 7.1 would be great however as Filip wrote web is still hosted on PHP 7.0 so we cannot merge this until we get better hosting.
Q: Is there a standard for `$i++`? Noticed changes `$i++` to `++$i`.

filipsuk commented:
@TomasVotruba @jakubzaba I successfully migrated eventigo.cz to PHP 7.1! 🎉 Tomas, please have a look at my comments (shown as outdated). It would be nice to merge this :)

TomasVotruba commented:
I though I'd have to downgrade, so this is great news :tada: 

@jakubzaba I think goal is to have ++one way to increment a value**. I respect that, since there is already a checker for it.

- `$i += 1`;
- `$i++`;
- `$i = $i + 1;` 
- `++$i`

Ther is [origin comment for it](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/831#issuecomment-87388273) to get more insight. 

Would you prefer another form for any reason? 

@filipsuk Wow! I asked Martin just week ago, was he so fast?

TomasVotruba commented:
I'll look on this.

TomasVotruba commented:
@filipsuk I've addressed your comments. Please check last commits, if that's all right.

The last thing I need help with is Sendgrid and category: https://github.com/eventigo/eventigo-web/pull/17/commits/5d7e3b14bf04e4a611b7950294e5f1e39aad260f#r114543965

filipsuk commented:
@TomasVotruba Good job, I will have a look at SendGrid. I see you converted whole repo to spaces, this should also be added to .editorconfig, right? I really like this PR 👍

filipsuk commented:
And also, will Travis do all this checking (including formatting)? Just asking if travis config needs to be updated

TomasVotruba commented:
Editorconfig update

> And also, will Travis do all this checking (including formatting)? Just asking if travis config needs to be updated

Yep, it is [on this line](https://github.com/eventigo/eventigo-web/blob/0cb56285b9369193039ca697e6425c59764c8a61/.travis.yml#L14).

Which is just [short cut in composer](https://github.com/eventigo/eventigo-web/blob/0cb56285b9369193039ca697e6425c59764c8a61/composer.json#L52).

So you can call it manually on your PC like:

```bash
composer cs # the shortes, for lazy like me :) 
# or
composer check-cs
# or 
vendor/bin/easy-coding-standard check app
```

Let me know, if it works on Windows. There might be some adjustments required.

TomasVotruba commented:
Ready to merge otherwise, feel free to do so.

TomasVotruba commented:
Thank you @filipsuk 

Let me know if you will have some ideas how to improve this and of this is pain in any way.
I realize standards should work for the developer, not the other way.
